geospaas_harvesting/harvesters.py

```diff
@@ -83,24 +83,64 @@ def harvest(self):
                 break
 
 
-class PODAACHarvester(Harvester):
-    """Harvester class for PODAAC data (NASA)"""
+class WebDirectoryHarvester(Harvester):
+    """
+    class for harvesting online data sources that rely on webpages (and most of the time on opendap)
+
+    Child classes should also assign values to the ingester and crawler class attributes.
+    These values should be the crawler and ingester classes to be used in the _create_crawlers()
+    and _create_ingester() methods.
+    """
+    ingester = None
+    crawler = None
+
     def _create_crawlers(self):
-        return [
-            crawlers.OpenDAPCrawler(url, time_range=(self.get_time_range()))
-            for url in self.config['urls']
-        ]
+        if self.crawler is None:
+            raise HarvesterConfigurationError(
+                "The class of crawler has not been specified properly")
+        try:
+            return [
+                self.crawler(url, time_range=(self.get_time_range()))
+                for url in self.config['urls']
+            ]
+        except (KeyError, TypeError, AttributeError,) as error:
+            raise HarvesterConfigurationError(
+                "crawler must be created properly with correct configuration file") from error
 
     def _create_ingester(self):
         parameters = {}
-        for parameter_name in ['max_fetcher_threads', 'max_db_threads']:
-            if parameter_name in self.config:
-                parameters[parameter_name] = self.config[parameter_name]
-        return ingesters.DDXIngester(**parameters)
+        if self.ingester is None:
+            raise HarvesterConfigurationError("The class of ingester has not been specified properly")
+        try:
+            for parameter_name in ['max_fetcher_threads', 'max_db_threads']:
+                if parameter_name in self.config:
+                    parameters[parameter_name] = self.config[parameter_name]
+            return self.ingester(**parameters)
+        except (KeyError, TypeError, AttributeError,) as error:
+            raise HarvesterConfigurationError(
+                "ingester must be created properly with correct configuration file") from error
 
 
+class PODAACHarvester(WebDirectoryHarvester):
+    """Harvester class for PODAAC data (NASA)"""
+    ingester = ingesters.DDXIngester
+    crawler = crawlers.OpenDAPCrawler
+
+
+class OSISAFHarvester(WebDirectoryHarvester):
+    """Harvester class for OSISAF project"""
+    ingester = ingesters.DDXIngester
+    crawler = crawlers.ThreddsCrawler
+    def __init__(self, **config):
+        super().__init__(**config)
+        if type(config['excludes']) != list:
+            raise HarvesterConfigurationError(
+                "'excludes' field must be fed with a list of excluded names ")
+        self.crawler.EXCLUDE.extend(config['excludes'])
```

Comment:
This kind of functionality can be useful for any `WebDirectoryHarvester`, so it should be handled in the `WebDirectoryHarvester._create_crawlers()`.

It should also be:
- an optional argument to the constructor of `WebDirectoryCrawler`
- stored in an instance attribute because when you modify the `EXCLUDE` class attribute from `Crawler`, you modify the value for all instances of all the child classes of `Crawler` (because `EXCLUDE` is a list, which is [mutable](https://medium.com/@meghamohan/mutable-and-immutable-side-of-python-c2145cf72747))

Reply:
Done in the next commit